Add `NoDatabasesReallocated` notification (#75)
neo4j/docs-status-codes: modules/ROOT/pages/notifications/all-notifications.adoc

Topology notifications provide additional information related to managing databases and servers.

[#_neo_clientnotification_cluster_nodatabasesreallocated]
=== NoDatabasesReallocated

.Notification details
[cols="<1s,<4"]
|===
|Code
m|Neo.ClientNotification.Cluster.NoDatabasesReallocated
|Title
a| `<command>` has no effect.
|Severity
m|INFORMATION
|Category
m|TOPOLOGY
|===

.Reallocating databases resulted in no allocation changes 
====
Command::
+
[source, cypher]
----
REALLOCATE DATABASES
----

Description of the returned code::

> No databases were reallocated.

Reviewer: Do we want to provide some further information, e.g. around that this is because the allocation already is stable/have converged (not sure what the correct term is)? Because as a user I would like to know if I e.g. should try again or if it was something that went wrong or if it is optimal already

Reviewer 2: I agree with Louise. I think it would be very helpful with one or more examples of when this can happen and how to "solve it".

Author: @Lojjs Sure, I think that would be a good idea. I've added "No better allocation is currently possible." Hopefully this indicates that nothing went wrong and if they tried again then they'd get the same outcome. I'm inclined to stay clear of saying things are "optimal" as it probably a bit of a loaded term/require too much explaining 😁 

@LinneaAndersson definitely I agree, I've added 2 examples. The first is where databases are allocated to all servers so there is not much a reallocate could do here. And the second when things look like they could be "improved" but there are constraints on the servers that prevent databases from moving so you're already in the "best" allocation.

Author: Also, I put the 5.13 label on this but I'm aware the cut is tomorrow. There is no pressure to get this in for then, so happy to bump it to 5.14 :)

Topology notifications provide additional information related to managing databases and servers.

[#_neo_clientnotification_cluster_nodatabasesreallocated]
=== NoDatabasesReallocated

.Notification details
[cols="<1s,<4"]
|===
|Code
m|Neo.ClientNotification.Cluster.NoDatabasesReallocated
|Title
a| `<command>` has no effect.
|Severity
m|INFORMATION
|Category
m|TOPOLOGY
|===

.Reallocating databases resulted in no allocation changes 
====
Command::
+
[source, cypher]
----
REALLOCATE DATABASES
----

Description of the returned code::
No databases were reallocated. No better allocation is currently possible. 

> To illustrate when the above code would be returned consider the case when there are three servers each hosting databases `foo` and `bar`, that is, all databases are allocated to all servers. In such a situation the allocation in the cluster is already balanced. 

Reviewer 3: ```suggestion

Example scenarios::
**Scenario 1:** The cluster is already balanced. 
For example, when there are three servers, each hosting databases `foo` and `bar`, meaning all databases are allocated to all servers.
+
```

> Another example of when this notification will be surfaced is when the cluster appears unbalanced but server constraints prevent us moving to a better, more balanced, allocation. For example, assuming server 1 hosts databases `foo` and `bar`, server 2 hosts only `foo`, and server 3 hosts no databases. Then a better allocation would move `foo` from server 1 to server 2, but if server 3 has the constraint `deniedDatabases:['foo']}`, then the cluster is already balanced subject to this constraint.

Author: Correct 😅

> Another example of when this notification will be surfaced is when the cluster appears unbalanced but server constraints prevent us moving to a better, more balanced, allocation. For example, assuming server 1 hosts databases `foo` and `bar`, server 2 hosts only `foo`, and server 3 hosts no databases. Then a better allocation would move `foo` from server 1 to server 3, but if server 3 has the constraint `deniedDatabases:['foo']}`, then the cluster is already balanced subject to this constraint.

Reviewer 3: ```suggestion
**Scenario 2:** The cluster appears unbalanced, but server constraints prevent you from moving to a better, more balanced, allocation. For example, assuming server 1 hosts databases `foo` and `bar`, server 2 hosts only `foo`, and server 3 hosts no databases. Then, a better allocation would move `foo` from server 1 to server 3, but if server 3 has the constraint `deniedDatabases:['foo']}`, then the cluster is already balanced subject to this constraint.
```
